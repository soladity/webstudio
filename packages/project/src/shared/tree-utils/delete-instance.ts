import { type Instance } from "@webstudio-is/react-sdk";

export const deleteInstanceMutable = (
  instance: Instance,
  instanceId: Instance["id"]
): boolean => {
  if (instance.id === instanceId) {
    return true;
  }

  for (const child of instance.children) {
    if (typeof child === "string") {
      continue;
    }
    const shouldDelete = deleteInstanceMutable(child, instanceId);
    if (shouldDelete === true) {
      const index = instance.children.indexOf(child);
      instance.children.splice(index, 1);
      return false;
    }
  }
  return false;
};
