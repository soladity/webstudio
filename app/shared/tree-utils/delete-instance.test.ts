import { type Instance } from "@webstudio-is/sdk";
import { deleteInstanceMutable } from "./delete-instance";

describe("Delete instance", () => {
  test("div 1 > div 2 > [div 3]", () => {
    const instance: Instance = {
      component: "Box",
      id: "1",
      style: {},
      children: [
        {
          component: "Box",
          id: "2",
          style: {},
          children: [
            {
              component: "Box",
              id: "3",
              style: {},
              children: [],
            },
          ],
        },
      ],
    };

    deleteInstanceMutable(instance, "3");
    expect(instance).toMatchSnapshot();
  });

  test("div 1 > div 2, [div 3]", () => {
    const instance: Instance = {
      component: "Box",
      id: "1",
      style: {},
      children: [
        {
          component: "Box",
          id: "2",
          style: {},
          children: [],
        },
        {
          component: "Box",
          id: "3",
          style: {},
          children: [],
        },
      ],
    };

    deleteInstanceMutable(instance, "3");
    expect(instance).toMatchSnapshot();
  });
});
