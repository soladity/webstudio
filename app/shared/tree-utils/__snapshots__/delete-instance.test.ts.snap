// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delete instance div 1 > div 2 > [div 3] 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "component": "Box",
      "id": "2",
      "style": Object {},
    },
  ],
  "component": "Box",
  "id": "1",
  "style": Object {},
}
`;

exports[`Delete instance div 1 > div 2, [div 3] 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "component": "Box",
      "id": "2",
      "style": Object {},
    },
  ],
  "component": "Box",
  "id": "1",
  "style": Object {},
}
`;
